Add a system to convert jsons between different MC versions

InfinityRaider opened the pull request:
I would like a review of this before merging though.

InfinityRaider committed:
Start a system to convert jsons between versions

InfinityRaider committed:
refactor

InfinityRaider committed:
Create version Convertors and clean up constructors

Also add class for weeds

InfinityRaider committed:
refactor

InfinityRaider committed:
Autoconvert old OreDict keys to tags

InfinityRaider committed:
Remove automatic en_US translation

InfinityRaider committed:
Make ResourcePlants not cloneable by default

InfinityRaider committed:
Auto Oredict on Products as well

InfinityRaider committed:
Seed Items should be items, not blocks

InfinityRaider committed:
Fix test

InfinityRaider committed:
Auto fix texture directories

RlonRyan commented:
Looks like something's up with your `git config user.email` setting or something and github isn't picking up that it's you doing the commits. Not that it matters.

RlonRyan commented on src/main/java/com/agricraft/agricore/json/AgriJsonVersion.java:
```diff
@@ -0,0 +1,21 @@
+import javax.annotation.Nullable;
+import java.util.function.Function;
+
+public interface AgriJsonVersion<T extends AgriSerializable> {
```
I should have added this to begin with.

InfinityRaider replied:
Well you did a great job with the initial implementation overall, and futureproofing something is quite hard...

RlonRyan commented on src/main/java/com/agricraft/agricore/json/AgriSerializable.java:
```diff
@@ -12,4 +12,6 @@
 
     void setPath(String path);
 
+    String getVersion();
```
This might be interesting if the AgriSerializables are nested, but I can't remember if they ever are.

RlonRyan replied:
Interesting in that you'd see the version number twice.

InfinityRaider replied:
They are not, there are nested classes, but none of the classes being nested are AgriSerializables

RlonRyan replied:
Ok good.

RlonRyan commented on src/main/java/com/agricraft/agricore/lang/AgriString.java:
```diff
@@ -15,20 +15,17 @@
         this.translations = translations;
         if (this.translations.isEmpty()) {
-            this.translations.put("en_US", normal);
         }
```
I'm not sure I understand this removal? I think in my original design the normal string is what is displayed if the translation fails. Was this something that was added in that last pull request?

InfinityRaider replied:
I did not like this to be honest, firstly, if I understand correctly, if the translation fails, it will return this either way, secondly, Minecraft has changed the language keys to be all lower case as well, and I do not feel like updating all jsons whenever mc changes the keys.

Furthermore, I am considering dropping this localization all together, since it can be perfectly done with resource packs as well.

RlonRyan replied:
How do the resource pack translations work? Do you just replace all the strings with keys?

RlonRyan commented on src/main/java/com/agricraft/agricore/plant/AgriPlant.java:
```diff
@@ -40,34 +42,29 @@
+        this.path = "default/weed_plant.json";
+        this.version = Versions_1_16.VERSION;
+        this.id = "weed_plant";
+        this.plant_name = new AgriString("Weed");
```
Does this need to be changed to something like unknown now that weeds are no longer "real" plants?

InfinityRaider replied:
I thought it'd be funny to leave it that way, given that weeds are unwanted, and we do not want jsons which fail to load either.

RlonRyan replied:
Sound logic. No turning back now.

RlonRyan commented on src/main/java/com/agricraft/agricore/registry/AgriSoils.java:
```diff
@@ -51,8 +53,8 @@ public boolean acceptsElement(String filename) {
-    public Class<AgriSoil> getElementClass() {
-        return AgriSoil.class;
+    public AgriJsonVersion<AgriSoil> getElementVersion() {
+        return Versions_1_16.SOIL;
```
I like this.

InfinityRaider replied:
I'm sure something will pop up sooner or later that will break this...

RlonRyan reviewed (approved):
Looks good to me!

💯

InfinityRaider committed:
Make json loading version aware

 - AgriCore will parse jsons against the version they are defined, and convert to the required version accordingly.
Not defining a conversion between versions will assume that the newer version can perfectly parse the old version without issues.

InfinityRaider committed:
Split reading and parsing of the json to avoid reading from file twice

InfinityRaider committed:
Annotations, typos and generics
